File: implement-shell-tools/cat/cat.js
import { program } from "commander";
import { promises as fs } from "node:fs";
import readline from "node:readline";
import process from "node:process";

program
  .name("concatenate-and-print-files-reproduction")
  .description("Print file content in the stdout.")
  .option("-n", "Number the output lines, starting at 1.")
  .option("-b", "Number the non-blank output lines, starting at 1.")
  .argument("[path...]", "the file path to process");

program.parse();

const argv = program.args;
const options = program.opts();

let rl;

start();

function start() {
  console.log(options.n);

  if (argv.length === 0) {
    rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    recursiveReadlineAndPrint();
  } else {
    readAndPrintFiles();
  }
}

function recursiveReadlineAndPrint() {
  rl.on("SIGINT", () => {
    rl.close();
    process.exit(0);
  }); 
  rl.question("", input => {
    console.log(input);
    recursiveReadlineAndPrint();
  });
}

async function readAndPrintFiles() {
  const buffer = [];
  for (const path of argv) {
    buffer.push(await readFile(path));
  }
  for(const content of buffer) {
    if (content[0].startsWith("cat.js:")) {
      console.error(content[0]);
    } else {
      let n = 1; 
      for(const string of content) {
        if (options.n) {
          console.log(`${n.toString().padStart(6, " ")}  ${string}`);
          n++;
        } else if (options.b) {
          if (string) {
            console.log(`${n.toString().padStart(6, " ")}  ${string}`);
            n++;
          } else {
            console.log();
          }
        } else {
          console.log(string);
        }
      }
    }
  }
}

async function readFile(path) {
  try {
    const content = await fs.readFile(path, "utf8");
    return content.toString().trim().split("\n");
  } catch (error) {
    return [`cat.js: ${path}: No such file or directory`];
  }
}

File: implement-shell-tools/cat/cat.py
import sys
import argparse


parser = argparse.ArgumentParser(
  prog = "concatenate-and-print-files",
  description = "The cat utility reads files sequentially, writing them to the standard output."
)

parser.add_argument("path", help = "The file to read and print", nargs = "*")
parser.add_argument("-b", action = "store_true", help = "Number the non-blank output lines, starting at 1.")
parser.add_argument("-n", action = "store_true", help = "Number the output lines, starting at 1.")

args = parser.parse_args()

def start():
  if (args.path):
    readAndPrintFile()
  else:
    try:
      while (True):
        string = input()
        print(string)
    except EOFError:
      pass
    except KeyboardInterrupt:
      pass

def readAndPrintFile():
  buffer = []
  for path in args.path:
    buffer.append(readFile(path))
  printBuffer(buffer)

def readFile(path):
  try:
    with open(path, "r") as file:
      content = file.read()
    return content.strip().split("\n")
  except:
    return  [f"cat.py: {path}: No such file or directory"]
  
def printBuffer(buffer):
  for fileContent in buffer:
    printContent(fileContent)

def printContent(fileContent):
  if fileContent[0].startswith("cat.py:"):
      print(fileContent[0], file = sys.stderr)
  elif (args.b):
    printWithFlagB(fileContent)
  elif(args.n):
    printWithFlagN(fileContent)
  else:
    printWithoutFlags(fileContent)

def printWithFlagB(fileContent):
  n = 1
  for string in fileContent:
    if (not string):
      print(string)
    else:
      print(str(n).rjust(6) + " " + string)
      n += 1

def printWithFlagN(fileContent):
  n = 1
  for string in fileContent:
    print(str(n).rjust(6) + " " + string)
    n += 1

def printWithoutFlags(fileContent):
  for string in fileContent:
    print(string)

start()
